src/leo_redundant_manager_worker.erl

```diff
@@ -730,39 +735,72 @@ redundancies_2(TableInfo, NumOfReplicas, L2, Members, VNodeId_0, R) ->
+                            redundancies_3(TableInfo, NumOfReplicas, NumOfRAs, Members, VNodeId, R);
                         _ ->
-                            redundancies_2(TableInfo, NumOfReplicas-1, L2, Members, VNodeId,
-                                           R#redundancies{temp_nodes = [Node_2|AccTempNode],
-                                                          temp_level_2 = AccLevel2_1,
-                                                          nodes = [#redundant_node{node = Node_2}
-                                                                   |AccNodes]})
+                            RackDict =
+                                lists:foldl(
+                                  fun(#redundant_node{node = N}, D) ->
+                                          case lists:keyfind(N, 1, Members) of
+                                              false ->
+                                                  D;
+                                              {_, RackId} ->
+                                                  dict:append(RackId, N, D)
+                                          end
+                                  end, dict:new(), AccNodes),
+                            RackL = dict:to_list(RackDict),
+
+                            case lists:keyfind(Node_2, 1, Members) of
+                                false ->
+                                    {error, []};
+                                {_, RackId} ->
+                                    case lists:keyfind(RackId, 1, RackL) of
+                                        {_,_} when length(RackL) /= NumOfRAs andalso
+                                                   NumOfReplicas - NumOfRAs < 1 ->
+                                            redundancies_3(
+                                              TableInfo, NumOfReplicas, NumOfRAs, Members, VNodeId, R);
+                                        %% Note: maximum total number of asigned racks
+                                        %%       adheres to `consistency.rack_aware_replicas`
+                                        %%       in `leo_manager.conf`
+                                        false when length(RackL) == NumOfRAs ->
+                                            redundancies_3(
+                                              TableInfo, NumOfReplicas, NumOfRAs, Members, VNodeId, R);
+                                        _ ->
+                                            redundancies_3(
+                                              TableInfo, NumOfReplicas - 1, NumOfRAs, Members, VNodeId,
+                                              R#redundancies{
+                                                temp_nodes = [Node_2|AccTempNode],
+                                                temp_level_2 = AccLevel2_1,
+                                                nodes = [#redundant_node{node = Node_2} | AccNodes]})
+                                    end
+                            end
```

Comment:
In this code block, we can eliminate

- generating RackL
  - Because we've already had the unique RackId list in temp_level_2 (before: AccLevel2  and after: AccLevel2_1)
- lists:keyfind(Node_2, 1, Members)
  - Because the same logic can be written by length(AccLevel2) == length(AccLevel2_1)

That being said, we can make the code more shorter and more efficient (less CPU cost) like the patch below

```diff
diff --git a/src/leo_redundant_manager_worker.erl b/src/leo_redundant_manager_worker.erl
index 23828ef..3adb327 100644
--- a/src/leo_redundant_manager_worker.erl                                                                                                                                                                  [0/784]
+++ b/src/leo_redundant_manager_worker.erl
@@ -759,47 +759,36 @@ redundancies_4(TableInfo, NumOfReplicas, NumOfRAs, Members, VNodeId, Node_1, R)
                 not_found ->
                     {error, ?ERROR_COULD_NOT_GET_REDUNDANCIES};
                 {Node_2, AccLevel2_1} ->
-                    LenAccLevel2 = length(AccLevel2_1),
+                    LenAccLevel2Before = length(AccLevel2),
+                    LenAccLevel2After = length(AccLevel2_1),

                     case (NumOfRAs /= 0 andalso NumOfRAs == length(AccNodes)) of
-                        true when LenAccLevel2 < NumOfRAs ->
+                        true when LenAccLevel2After < NumOfRAs ->
                             redundancies_3(TableInfo, NumOfReplicas, NumOfRAs, Members, VNodeId, R);
                         _ ->
-                            RackDict =
-                                lists:foldl(
-                                  fun(#redundant_node{node = N}, D) ->
-                                          case lists:keyfind(N, 1, Members) of
-                                              false ->
-                                                  D;
-                                              {_, RackId} ->
-                                                  dict:append(RackId, N, D)
-                                          end
-                                  end, dict:new(), AccNodes),
-                            RackL = dict:to_list(RackDict),
-
-                            case lists:keyfind(Node_2, 1, Members) of
-                                false ->
-                                    {error, []};
-                                {_, RackId} ->
-                                    case lists:keyfind(RackId, 1, RackL) of
-                                        {_,_} when length(RackL) /= NumOfRAs andalso
-                                                   NumOfReplicas - NumOfRAs < 1 ->
-                                            redundancies_3(
-                                              TableInfo, NumOfReplicas, NumOfRAs, Members, VNodeId, R);
-                                        %% Note: maximum total number of asigned racks
-                                        %%       adheres to `consistency.rack_aware_replicas`
-                                        %%       in `leo_manager.conf`
-                                        false when length(RackL) == NumOfRAs ->
-                                            redundancies_3(
-                                              TableInfo, NumOfReplicas, NumOfRAs, Members, VNodeId, R);
-                                        _ ->
-                                            redundancies_3(
-                                              TableInfo, NumOfReplicas - 1, NumOfRAs, Members, VNodeId,
-                                              R#redundancies{
-                                                temp_nodes = [Node_2|AccTempNode],
-                                                temp_level_2 = AccLevel2_1,
-                                                nodes = [#redundant_node{node = Node_2} | AccNodes]})
-                                    end
+                            % What we have to be able to check in order to ensure the constraints of the rack awareness
+                            % 1. How many racks/which RackId we've already assigned
+                            %   - We can reference this by checking LenAccLevel2Before or checking AccLevel2 itself
+                            % 2. Is the value of 1 changed(found a new one) after invoking get_redundancies?
+                            %   - This can be confirmed by comparing LenAccLevel2Before and LenAccLevel2After
+                            case LenAccLevel2Before == LenAccLevel2After of
+                                true when LenAccLevel2Before /= NumOfRAs andalso
+                                          NumOfReplicas - NumOfRAs < 1 ->
+                                    redundancies_3(
+                                      TableInfo, NumOfReplicas, NumOfRAs, Members, VNodeId, R);
+                                %% Note: maximum total number of asigned racks
+                                %%       adheres to `consistency.rack_aware_replicas`
+                                %%       in `leo_manager.conf`
+                                false when LenAccLevel2Before == NumOfRAs ->
+                                    redundancies_3(
+                                      TableInfo, NumOfReplicas, NumOfRAs, Members, VNodeId, R);
+                                _ ->
+                                    redundancies_3(
+                                      TableInfo, NumOfReplicas - 1, NumOfRAs, Members, VNodeId,
+                                      R#redundancies{
+                                        temp_nodes = [Node_2|AccTempNode],
+                                        temp_level_2 = AccLevel2_1,
+                                        nodes = [#redundant_node{node = Node_2} | AccNodes]})
                             end
                     end
             end
```

I confirmed that the unit tests with this patch applied have passed several times on my dev-box.